runtime/decimal.c

```diff
@@ -45,21 +45,57 @@ TaggedPtrPanicCode _bal_decimal_sub(TaggedPtr tp1, TaggedPtr tp2) {
     return finish(&d, &cx);    
 }
 
+TaggedPtrPanicCode _bal_decimal_mul(TaggedPtr tp1, TaggedPtr tp2) {
+    const decQuad *dq1 = taggedToDecQuad(tp1);
+    const decQuad *dq2 = taggedToDecQuad(tp2);
+    decQuad d;
+    decContext cx;
+    initContext(&cx);   
+    decQuadMultiply(&d, dq1, dq2, &cx);
+    return finish(&d, &cx); 
+}
+
+TaggedPtrPanicCode _bal_decimal_div(TaggedPtr tp1, TaggedPtr tp2) {
+    const decQuad *dq1 = taggedToDecQuad(tp1);
+    const decQuad *dq2 = taggedToDecQuad(tp2);
+    decQuad d;
+    decContext cx;
+    initContext(&cx);   
+    decQuadDivide(&d, dq1, dq2, &cx);
+    return finish(&d, &cx); 
+}
+
 TaggedPtrPanicCode finish(decQuad *dq, decContext *cx) {
     TaggedPtrPanicCode result;
+    enum decClass class = decQuadClass(dq);
+    if (class == DEC_CLASS_POS_ZERO || 
+        class == DEC_CLASS_NEG_ZERO || 
+        class == DEC_CLASS_POS_SUBNORMAL || 
+        class == DEC_CLASS_NEG_SUBNORMAL) {
+        decQuadZero(dq);
+        result.panicCode = 0;
+        result.ptr = createDecimal(dq);
+        return result;
+    }
     uint32_t status = cx->status;
     if (status & DECIMAL_STATUS_FAIL) {
-        // XXX decide what panic codes to use 
-        result.panicCode = (status & DEC_Division_by_zero) ? PANIC_DIVIDE_BY_ZERO : PANIC_ARITHMETIC_OVERFLOW;
+        if (status & DEC_Division_by_zero) {
+            result.panicCode = PANIC_DIVIDE_BY_ZERO;
+        }
+        else if (status & DEC_Underflow) {
```

Comment:
Why are you doing this? What are the values in this case? Won't the class be subnormal?

Reply:
Please consider this example:
`mul(9.999999999999999999999999999999999E-6001, 1E-143)`
In decQuad module, the class of the result is +Normal and value of the result is `1.000000000000000000000000000000000E-6143` because it is rounded. Also, undeflowed and inexact flags are raised.

But in subset, the class of the result is +Zero and the value of the result is `0`. Also, undeflowed, subnormal, inexact and rounded flags are raised. 

In subset, subnormal values are not permited, underflow flag is raised and value becomes zero. 
In other case, rounding was applies to the result(9.999... -> 10.000...) which is a subnormal and value becomes +Normal (9.999...E-6144 -> 10.000...E-6144 ->1.000...E-6143).

Reply:
Let me see if I understand.  You are saying:

1. Multiplication happens in two logical stages: perform the multiplication to a precision greater than allowed by decimal128 format, and then round to fit into a decimal128
2. It is possible for the result of the first stage to be subnormal, but then be rounded in the second stage to become normal
3. In this case, decNumber in subset mode will return 0.
4. We can detect this case in decQuad by checking the Underflow case

Is that right?

If so, that's a very interesting case.  I am glad you have discovered it.

But I don't think the Ballerina spec requires us to return 0. Not allowing subnormal values means that there are no Ballerina decimal values that are subnormal. It does not constrain internal intermediate results.   My concern is that this may be very difficult to implement in jBallerina.  Can this be done on top of BigDecimal?

Reply:
> Is that right?
> 

Yes, that is correct.

> But I don't think the Ballerina spec requires us to return 0. Not allowing subnormal values means that there are no Ballerina decimal values that are subnormal. It does not constrain internal intermediate results. My concern is that this may be very difficult to implement in jBallerina. Can this be done on top of BigDecimal?

@jclark, I have tried it here https://github.com/KavinduZoysa/decimal/blob/master/decimal.java. I am not sure whether this is the correct approach doing this using BigDecimal. In BigDecimal, setting context to DECIMAL128 only limit the number of digits to 34 and does not limit the value of exponent. So I think we can compare the result with minimum magnitute to detect the subnormal case.

Reply:
That will, presumably, give the same result as testing whether class is subnormal. How can it handle the case where the intermediate result (before rounding) is subnormal but the result after rounding is normal?